### PR TITLE
[UNR-2359] Spatial Debugger Config UI

### PR DESCRIPTION
**Pull request by KeyMaster-**
#### Description
spatial_debugger_config_ui_preview1
Adds an in-game UI to change spatial debugger settings.
Currently this covers only flags relating to tag drawing (except the draw range) and worker regions drawing.
Currently enabled via F8, this still needs to be made configurable.

#### Release note
TODO
REQUIRED: Add a release note to the `##Unreleased` section of CHANGELOG.md. You can find guidance for writing useful release notes [here](../SpatialGDK/Extras/internal-documentation/how-to-write-good-release-notes.md). Documentation changes are exempt from this requirement.

#### Tests
Manual, no automated tests

STRONGLY SUGGESTED: How can this be verified by QA?

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?
TODO

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 1]
```diff
@@ -282,7 +284,65 @@ void ASpatialDebugger::OnEntityAdded(const Worker_EntityId EntityId)
 		if (Actor->IsA<APlayerController>())
 		{
 			LocalPlayerController = Cast<APlayerController>(Actor);
+
+			if (ConfigUIWidget == nullptr)
+			{
+				FSoftClassPath ConfigUIClassRef(DEFAULT_CONFIG_UI_WIDGET_CLASS);
+				UClass* ConfigUIClass = ConfigUIClassRef.TryLoadClass<UUserWidget>();
+				if (ConfigUIClass)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:292]:
```suggestion
				if (ConfigUIClass != nullptr)
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 2]
```diff
@@ -282,7 +284,65 @@ void ASpatialDebugger::OnEntityAdded(const Worker_EntityId EntityId)
...
+				{
+					ConfigUIWidget = CreateWidget<USpatialDebuggerConfigUI>(LocalPlayerController.Get(), ConfigUIClass);
+					ConfigUIWidget->SetSpatialDebugger(this);
+				}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:296]: Should we print a warning if either the widget is unset, or this TryLoadClass fails?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 3]
```diff
@@ -282,7 +284,65 @@ void ASpatialDebugger::OnEntityAdded(const Worker_EntityId EntityId)
...
+	if (ConfigUIWidget == nullptr)
+	{
+		return;
+	}
+
+	if (!ConfigUIWidget->IsInViewport())
+	{
+		ConfigUIWidget->AddToViewport();
+
+		// TODO This is clobbering whatever input mode the player controller had set before we enabled our UI.
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:314]: I think we should store the old one in a var 👍

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebuggerConfigUI.cpp [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Utils/SpatialDebuggerConfigUI.h"
+#include "EngineClasses/SpatialNetDriver.h"
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebuggerConfigUI.cpp:4]: Put a line break between these two, we always want the .h first, and I think clang-format might reorder these two

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h [hunk 5]
```diff
@@ -48,6 +49,17 @@ struct FWorkerRegionInfo
...
+UENUM()
+namespace EActorTagDrawMode
+{
+enum Type
+{
+	None,
+	LocalPlayer,
+	All
+};
+}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:61]: I think running clang-format should do this but
```suggestion
} // namespace EActorTagDrawMode
```

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h [hunk 6]
```diff
@@ -67,6 +79,13 @@ class SPATIALGDK_API ASpatialDebugger : public AInfo
 	UFUNCTION(Exec, Category = "SpatialGDK", BlueprintCallable)
 	void SpatialToggleDebugger();
 
+	UFUNCTION(Category = "SpatialGDK", BlueprintCallable, BlueprintPure)
+	bool IsEnabled();
+
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = General,
+			  meta = (ToolTip = "Key to open configuration UI for the debugger at runtime"))
+	FKey ConfigUIKey;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:87]: Should this category be SpatialGDK?

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h [hunk 7]
```diff
@@ -78,6 +97,9 @@ class SPATIALGDK_API ASpatialDebugger : public AInfo
 			  meta = (ToolTip = "Maximum range from local player that tags will be drawn out to"))
 	float MaxRange = 100.0f * 100.0f; // 100m
 
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = Visualization, meta = (Tooltip = "Which actor tags to show"))
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:100]:
```suggestion
	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = Visualization, meta = (Tooltip = "Which Actor tags to show"))
```

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebuggerConfigUI.h [hunk 8]
```diff
@@ -0,0 +1,26 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "Blueprint/UserWidget.h"
+#include "CoreMinimal.h"
+#include "SpatialDebuggerConfigUI.generated.h"
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebuggerConfigUI.h:7]:
```suggestion

#include "SpatialDebuggerConfigUI.generated.h"
```

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs [hunk 9]
```diff
@@ -41,7 +41,10 @@ public SpatialGDK(ReadOnlyTargetRules Target) : base(Target)
                 "OnlineSubsystemUtils",
                 "Projects",
                 "ReplicationGraph",
-                "Sockets"
+                "Sockets",
+                "Slate",
+                "SlateCore",
+                "UMG"
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs:47]: Double-checking these are all needed?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebuggerConfigUI.cpp (hunk 4 above)
Comment (KeyMaster-) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebuggerConfigUI.cpp:4]: Turns out it doesn't! Still gonna add the linebreak

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h (hunk 5 above)
Comment (KeyMaster-) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialDebugger.h:61]: Doesn't seem to. Gotta do everything yourself around here!

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs (hunk 9 above)
Comment (KeyMaster-) [reply SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs:47]: Good call, turns out SlateCore is redundant when also including Slate (which is required)

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 3 above)
Comment (KeyMaster-) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:325]: I looked into it further, and the input mode doesn't get stored anywhere, meaning that we can't retrieve it before changing it. Applying an input mode also touches a bunch of state, and storing that state before applying the input mode seems difficult if not impossible.
So it doesn't look like showing this UI can be done completely invisibly to user code.
My best alternate idea would be a callback for user code to hook into and restore state as it needs to, but that'd require some additional code to make it easy for users to subscribe to that callback when the spatial debugger is created.

**Comment by alastairdglennie** [comment]
Also if this had an easy way to alternative the height and vertical scale of the worker regions that'd be ace. 

Currently, configuring the draw height is done by editing the `const float DEFAULT_WORKER_REGION_HEIGHT` in WorkerRegion.cpp and `Worker Region Vertical Scale` is a SpatialDebugger property.

Both of these ^ can be done live by select the Actors in the client world outliner and setting, but being able to do through this would be sick.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 10]
```diff
@@ -35,7 +35,9 @@ namespace
 {
```
[the remark this answers is not in the document]
Comment (KeyMaster-) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:35]: Gonna test this with 4.24 to see how it goes.

**Comment by KeyMaster-** [comment]
Ah I can add that, good call.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 11]
```diff
@@ -35,7 +35,7 @@ namespace
...
 const FString DEFAULT_WORKER_REGION_MATERIAL =
 	TEXT("/SpatialGDK/SpatialDebugger/Materials/TranslucentWorkerRegion.TranslucentWorkerRegion");
-}
+} // namespace
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:38]: Does running clang-format not do this automatically?
```suggestion
} // anonymous namespace
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 12]
```diff
@@ -282,10 +282,85 @@ void ASpatialDebugger::OnEntityAdded(const Worker_EntityId EntityId)
...
 			LocalPlayerController = Cast<APlayerController>(Actor);
+
+			if (!ConfigUIWidget.IsValid())
+			{
+				if (ConfigUIClass != nullptr)
+				{
+					ConfigUIWidget = CreateWidget<USpatialDebuggerConfigUI>(LocalPlayerController.Get(), ConfigUIClass);
+					if (!ConfigUIWidget.IsValid())
+					{
+						UE_LOG(LogSpatialDebugger, Error, TEXT("Couldn't create config UI widget."));
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:293]: maybe something along these lines?
```suggestion
						UE_LOG(LogSpatialDebugger, Error, TEXT("SpatialDebugger config UI will not load. Couldn't create config UI widget for class: %s"), GetNameSafe(*ConfigUIWidget));
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 11 above)
Comment (KeyMaster-) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:38]: Doesn't seem to. Didn't do it for the enum namespace either.

**Comment by KeyMaster-** [comment]
Exposing region height and opacity has been moved into its own PR here: https://github.com/spatialos/UnrealGDK/pull/2585

File: SpatialGDK/Source/SpatialGDK/Public/Utils/BasicAwaiter.h [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+UCLASS(Blueprintable, DefaultToInstanced)
+class SPATIALGDK_API UBasicAwaiter : public UObject, public IAwaitable
+{
+	GENERATED_BODY()
+
+public:
+	virtual ~UBasicAwaiter() {}
+
+	/** IAwaitable Implementation */
+	virtual FDelegateHandle Await(const FOnReady& OnReadyDelegate, const float Timeout = 0.f) override;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/BasicAwaiter.h:29]: Explicitly callout that timeout of zero means no timeout

**Comment by KeyMaster-** [comment]
Here are screenshots of some of the blueprint changes. However as this is a widget a lot of it is distributed over various functions or done via bindings. I'll do a screen-share review with someone soon.

2020-10-06 11_43_42-GDKTestGyms  Debug  - Unreal Editor
2020-10-06 11_43_34-GDKTestGyms  Debug  - Unreal Editor
2020-10-06 11_43_12-GDKTestGyms  Debug  - Unreal Editor
2020-10-06 11_38_52-GDKTestGyms  Debug  - Unreal Editor

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialBasicAwaiter.cpp [hunk 14]
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Utils/SpatialBasicAwaiter.h"
+
+#include "Engine/World.h"
+#include "TimerManager.h"
+
+const FString USpatialBasicAwaiter::TIMEOUT_MESSAGE = FString{ TEXT("Timed out while waiting to become Ready!") };
+constexpr TCHAR DESTROYED_MESSAGE_TEMPLATE[] = TEXT("Awaiter '%s' was destroyed!");
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialBasicAwaiter.cpp:9]: Just use this in place

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialBasicAwaiter.cpp [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+void USpatialBasicAwaiter::BeginDestroy()
+{
+	InvokeQueuedDelegates(FString::Printf(DESTROYED_MESSAGE_TEMPLATE, *GetName()));
+	Super::BeginDestroy();
+}
+
+FDelegateHandle USpatialBasicAwaiter::Await(const FOnReady& OnReadyDelegate, const float Timeout)
+{
+	if (bIsReady)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialBasicAwaiter.cpp:19]: Could you decorate a couple of comments here about what each path means.
ie. it's already ready, so no need to wait, execute immediately

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialBasicAwaiter.cpp [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+	{
+		return;
+	}
+
+	bIsReady = false;
+	OnResetEvent.Broadcast();
+	OnResetEvent.Clear();
+}
+
+void USpatialBasicAwaiter::InvokeQueuedDelegates(const FString& ErrorStatus /* = FString */)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialBasicAwaiter.cpp:93]:
```suggestion
void USpatialBasicAwaiter::InvokeQueuedDelegates(const FString& ErrorStatus /* = FString{} */)
```

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs [hunk 17]
```diff
@@ -41,7 +41,9 @@ public SpatialGDK(ReadOnlyTargetRules Target) : base(Target)
                 "OnlineSubsystemUtils",
                 "Projects",
                 "ReplicationGraph",
-                "Sockets"
+                "Sockets",
+                "Slate",
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs:45]: Need both slate and umg?

**Review by m-samiec — APPROVED** [review]
Nice work!

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs (hunk 17 above)
Comment (KeyMaster-) [reply SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs:45]: UMG for widget types, Slate for input types when binding the toggle key.

**Pull request merged**